### ISSUE TITLE
Roadmap: single-GPU sessions, explicit eager execution, and selective tracing

### ISSUE THREAD

**Issue by shinaoka** [tensor4all/tensor4all-rs#623]
## Summary

Migrate `tensor4all-rs` to the post-[tenferro-rs#1680](https://github.com/tensor4all/tenferro-rs/issues/1680) execution model in three stages:

1. keep the current ITensors.jl-like, backend-implicit eager API while adopting tenferro sessions internally and adding single-CUDA-GPU support;
2. make eager execution session-explicit in a planned breaking release;
3. add selective traced regions alongside eager execution, with native execution, StableHLO lowering, optional PJRT execution, and a later JAX FFI integration path.

This issue defines the migration architecture and sequencing. It also provides the execution-model foundation for the existing GPU request [#553](https://github.com/tensor4all/tensor4all-rs/issues/553).

## Motivation

Today `tensor4all-rs` hides backend/session management from users, similar to ITensors.jl. Internally, `tensor4all-tensorbackend` owns process-global CPU objects for plain operations, graph execution, and eager AD. This is convenient, but it prevents coherent CPU/CUDA placement, repeats session entry, and makes explicit execution control difficult.

Tenferro is moving to one canonical concrete API in which operations accept a borrowed `&mut dyn BackendSession` and the backend-taking one-shot API is removed. Tensor4all should adopt that model without combining all downstream breakage into one change.

The long-term design should also support tracing only selected numerical regions. Outer tensor-network control flow should remain ordinary eager Rust, while suitable static subcomputations can be compiled once and executed through the native runtime or the optional XLA/PJRT peer executor.

## Current constraints

- `tensor4all-tensorbackend/src/context.rs` currently owns separate process-global CPU backend, graph compiler/executor, and eager runtime objects.
- `TensorDynLen` may hold host `Storage` or a runtime-bound `EagerTensor`.
- `Storage` and `Matrix` are host/`Vec` based.
- Existing native einsum already uses `TracedTensor` internally, but traced values are not part of the tensor4all public API.
- Tenferro CUDA requires explicit upload/download and rejects incompatible host/device buffers instead of silently transferring them.
- Tenferro XLA is a peer whole-program executor over `CompiledGraph`/`SemanticProgram`; it is not a `TensorBackend`.
- CUDA full-pivot LU is not available, so MatrixLUCI/TCI must not be advertised as GPU-capable merely because CUDA feature plumbing exists.

## Design invariants

- No implicit CPU/CUDA transfer.
- No unsupported-operation fallback from CUDA, PJRT, or tracing to eager CPU execution.
- Tensor placement is derived from the underlying tenferro tensor; tensor4all must not add a second independent device tag.
- Only one CUDA device is supported initially: visible ordinal `0`. Physical selection may use `CUDA_VISIBLE_DEVICES`.
- CPU and CUDA caches/runtimes are separate and explicitly owned.
- A high-level operation enters at most one backend session; helpers receiving a session must not re-enter a backend.
- Full `Index` semantics are preserved, including prime level, tags, direction, and ordering. Index IDs alone are not sufficient.
- Eager and traced values are separate types, not variants of one runtime-mode enum.

## Stage 1: backend-implicit eager execution with one CUDA GPU

Preserve the existing tensor/network operation signatures. Users do not handle a backend or session.

Internally, route concrete execution through one placement-aware seam in `tensor4all-tensorbackend`, conceptually:

```text
existing tensor4all API
        |
validate input placement
        |
with_session_for(inputs, |session| ...)
        |-- host  -> CPU session
        `-- cuda0 -> CUDA session
```

Add optional CUDA feature plumbing for the upstream crates required by the supported vertical slice, including `tenferro-gpu/cuda`, `tenferro-ad/cuda`, `tenferro-einsum/cuda`, and `tenferro-linalg/cuda` where applicable after #1680 lands.

GPU use remains explicit through tensor placement, without exposing backend/session objects:

```rust
let a_gpu = a.upload_cuda()?;
let b_gpu = b.upload_cuda()?;
let c_gpu = a_gpu.contract(&b_gpu)?;
let c_cpu = c_gpu.download()?;
```

Requirements:

- ordinary constructors continue to create host tensors;
- `upload_cuda` targets visible CUDA ordinal `0` only;
- `download` is the explicit device-to-host boundary;
- results remain on the input device;
- mixed host/CUDA operands return a typed placement error;
- unavailable CUDA or unsupported dtype/op returns a typed error, never CPU fallback;
- tracked tensors are uploaded before enabling gradients; cross-runtime transfer of tracked values is initially rejected unless an explicit detach contract is used;
- host-only `Storage`/`Matrix` paths reject CUDA-resident values unless the public API explicitly documents a host readback boundary;
- unavoidable scalar/rank-selection readbacks in truncation algorithms are documented and measured.

The initial algorithm-level GPU vertical slice should use a supported TreeTN contraction or another tensor-network path that remains device-resident. TCI/ACI remain CPU-only until their required linalg operations and host `Matrix` representation have a separate GPU design.

## Stage 2: explicit eager sessions

At a planned breaking release, make the session the canonical eager execution argument. A backend is used to enter a bounded session; operations consume the already-entered session.

Conceptually:

```rust
backend.with_backend_session(|session| {
    let c = a.contract(&b, session)?;
    let d = c.add(&x, session)?;
    d.truncate(options, session)
})
```

Requirements:

- use one unsuffixed operation vocabulary;
- do not permanently retain parallel implicit and `*_in`/`*_with_session` APIs;
- remove process-global default execution from the canonical path;
- remove or privatize the CPU-specific `with_default_backend` surface;
- keep ordinary tensor values context-free where possible;
- separate ordinary concrete tensors from runtime-owned AD tracking rather than storing a session inside tensor values;
- make cross-session/device mismatches typed errors;
- give graph, plan, and backend caches explicit bounded owners with stats and clear/reset controls;
- update the C API/language bindings using opaque runtime/session handles rather than exposing Rust trait objects.

## Stage 3: selective trace regions alongside eager execution

Do not add a mode flag to `TensorDynLen`. Introduce a sibling symbolic type and explicit region/artifact types, conceptually:

```text
TensorDynLen              concrete eager value
TracedTensorDynLen        symbolic value inside one trace region
TraceRegion               tensor4all graph plus input/output metadata contract
CompiledRegion            tenferro CompiledGraph plus tensor4all bindings
```

Only the selected region is traced. Outer topology manipulation, sweeps, convergence checks, pivot/rank decisions, and data-dependent Rust control flow remain eager.

Conceptual use:

```rust
let region = TraceRegion::build(&input_specs, |inputs| {
    let y = inputs[0].contract(&inputs[1])?;
    Ok(vec![y])
})?;

let compiled = region.compile()?;
let outputs = native_executor.run(session, &compiled, &[&a, &b])?;
```

A trace region may be compiled for native execution, lowered to StableHLO, or prepared for optional PJRT execution through the following paths:

1. the native tenferro runtime using an explicit CPU/CUDA session;
2. StableHLO lowering for inspection/export;
3. the optional `tenferro-xla` PJRT executor.

XLA/PJRT must remain a peer whole-program executor, not a `TensorBackend` variant.

### Tensor4all metadata contract

`CompiledRegion` must retain tensor4all input/output metadata outside the numeric tenferro graph:

- ordered full indices and equality relationships;
- logical shapes and dtypes;
- layout and placement requirements;
- output index reconstruction;
- supported storage class.

Runtime binding validates full-index relationships, shape, dtype, layout, and placement. The numerical executable cache may be shape/dtype/layout specialized without keying on concrete Index UUIDs, while each region retains the binding contract needed to reconstruct output indices safely.

The first traced slice should avoid operations that create new tensor-network indices. A dense static contraction whose output indices are inherited from its inputs is sufficient to validate the boundary.

### Initial traced capability

Start with the static real-floating subset verified against the tenferro XLA/StableHLO implementation available when Stage 3 begins:

- dense `F32`/`F64`;
- add, multiply, negate;
- reshape, transpose, broadcast, convert;
- reduce-sum and dot-general;
- einsum/contraction lowered to those operations.

Reject with typed compile-time errors:

- structured storage requiring implicit densification;
- dynamic shapes;
- complex/integer/bool types until supported deliberately;
- data-dependent control flow;
- SVD/QR/eig/truncation;
- unsupported extension operations;
- eager fallback or graph breaks.

### AD and JAX scope

Tracing is execution staging, not AD. The first traced slice is forward-only. Later, tenferro graph AD may produce VJP/JVP graphs that are compiled and executed through the same native/PJRT paths.

Do not call StableHLO/PJRT execution “JAX support”. There are two distinct goals:

- using XLA from tensor4all: trace -> StableHLO -> PJRT;
- calling a tensor4all compiled region from a JAX program: a separate JAX FFI primitive with abstract evaluation, layout/batching behavior, and explicit JVP/VJP rules.

A JAX adapter should therefore be a follow-up within Stage 3 after native/PJRT compiled-region execution is stable.

## Suggested implementation order

1. Land tenferro-rs #1680.
2. Update the tenferro pin and migrate tensor4all internal CPU calls to the session-explicit tenferro API.
3. Centralize backend/session entry in `tensor4all-tensorbackend`.
4. Add CUDA feature plumbing, explicit upload/download, placement validation, and one supported GPU tensor-network benchmark for #553.
5. Introduce the breaking explicit-session eager API and remove implicit canonical execution.
6. Add `TracedTensorDynLen`, metadata-preserving region compilation, and native execution parity.
7. Add StableHLO lowering and optional PJRT execution for the supported static subset.
8. Add a separate JAX FFI adapter only after the compiled-region ABI and AD contracts are stable.

Each numbered migration stage may be split into multiple focused PRs. Do not combine dependency migration, CUDA support, the breaking public API, and traced execution in one PR.

### Stage 1

- CPU-only default feature build and full release-mode tests remain green.
- CUDA feature builds without affecting CPU-only users.
- CUDA ordinal `0` numerical parity and output-residency tests pass.
- Mixed placement and unsupported CUDA operations fail without fallback.
- No high-level operation opens nested backend sessions.
- At least one representative tensor-network CUDA benchmark reports setup, transfer, and steady-state execution separately.

### Stage 2

- Canonical eager operations require an explicit session.
- One session covers a mixed tensor/network operation chain.
- The old implicit canonical surface is removed at the breaking boundary.
- Runtime/cache ownership is explicit and bounded.
- CPU and CUDA parity, AD, C API, and language-binding migration tests pass.

### Stage 3

- Eager and traced execution produce matching numerical values and complete output Index metadata.
- One compiled region runs repeatedly with different compatible values.
- Invalid Index/shape/dtype/layout/placement bindings return typed errors.
- Unsupported tracing content returns a typed trace error. Content unsupported by the selected lowering or executor returns a typed compile/execution error; it is never silently redirected to eager CPU or another executor.
- Compile latency and steady-state execution are measured separately.
- Compiled/executable caches have bounded defaults, stats, and clear controls; keys cover graph fingerprint, shapes, dtypes, layout, target, lowering version, and runtime/plugin identity.
- Native execution parity passes for the supported subset. When the optional PJRT feature is enabled and the target runtime is available, PJRT/native parity also passes for the PJRT-supported subset.
- JAX integration is claimed only after `jax.jit` execution works; differentiation is claimed only after explicit JVP/VJP and finite-difference tests pass.

## Non-goals

- Multi-GPU execution, sharding, collectives, or device meshes.
- Automatic backend selection that moves tensors silently.
- CUDA fallback to CPU for unsupported operations.
- Tracing an entire dynamic tensor-network algorithm in the first trace milestone.
- Dynamic shapes or tensor-value-dependent control flow in the first trace milestone.
- Treating PJRT, StableHLO, and JAX as interchangeable concepts.
- Maintaining two permanent equivalent eager API vocabularies after the Stage 2 break.

## Related issues

- GPU request: [#553](https://github.com/tensor4all/tensor4all-rs/issues/553)
- Tenferro canonical session API: [tenferro-rs#1680](https://github.com/tensor4all/tenferro-rs/issues/1680)
- Tenferro session design background: [tenferro-rs#1673](https://github.com/tensor4all/tenferro-rs/issues/1673)
- Tenferro XLA/StableHLO/PJRT executor: [tenferro-rs#984](https://github.com/tensor4all/tenferro-rs/issues/984)

## Open questions

- Which TreeTN operation should be the first CUDA benchmark and supported vertical slice?
- Which Stage 1 public transfer names best communicate synchronization and AD detachment semantics?
- Should Stage 2 expose tenferro `BackendSession` directly or a narrow tensor4all session wrapper required by Index/structured-tensor operations?
- Which index-binding relations should be reusable across concrete Index identities in the first traced region API?
- Should the first PJRT milestone return host tensors only, or require an explicit output-placement policy from the start?

**Comment by shinaoka** [comment]
## Design amendment: multi-GPU dispatch and parallel execution requirements

Two upcoming requirements were not part of the original design and invalidate some Stage 1/2 assumptions. They should be incorporated now, before Stage 2 fixes the session API shape.

1. **Task parallelism via [hataori-rs](https://github.com/shinaoka/hataori-rs)**: MPI, Rayon, MPI+Rayon, and nested Rayon parallelism, with coarse tasks (seconds to minutes).
2. **Multi-GPU dispatch**: one process holds several GPUs and sends a whole computation to a chosen device. This is dispatch, not partitioned/distributed execution of a single contraction.

### Where the current design conflicts

**"Visible ordinal 0 only" cannot be the API shape.** Physical device selection via `CUDA_VISIBLE_DEVICES` assumes one device per process. With in-process dispatch, `CudaExecutionContext` must be constructible per device, e.g. `CudaExecutionContext::new_on(ordinal)`. The type is already a value with `context_id` validation, so it has the right shape; the single-device restriction should be restated as an initial implementation limit, not as an API assumption. Nothing may rely on implicit current-device thread state (`cudaSetDevice` style), because a Rayon worker thread will execute tasks bound to different contexts over its lifetime.

**The process-global CPU context serializes Rayon tasks.** `tensor4all-tensorbackend/src/context.rs` keeps `DEFAULT_CONTEXT: OnceLock<Arc<CpuExecutionContext>>`, whose `with_backend` holds a `Mutex<CpuBackend>` for the duration of the closure. Any parallelism funneled through the global default contends on that lock. (The eager path shares an `Arc<EagerRuntime>`; its internal lock granularity needs a separate check.) The fix is already in the Stage 2 requirement "remove process-global default execution from the canonical path", but the design should additionally state the ownership model explicitly: **one context is one serialization domain; parallelism is obtained by creating one context per task or per worker, never by sharing the global default across tasks.**

**Nested Rayon needs a thread-pool injection point.** When an outer hataori/Rayon task runs tensor4all code whose backend (e.g. faer) spawns inner Rayon work, oversubscription must be controllable from the outside. `CpuExecutionContext` should accept an optional `rayon::ThreadPool` (e.g. `CpuExecutionContext::with_thread_pool(...)`) so the caller can bound inner parallelism per task.

**Context-free constructors inside algorithms are a structural hole.** #720 found that SRC builds probes and caps via `T::from_dense` / `T::ones`, which land in the global default CPU context regardless of input placement. This is not an SRC-specific bug: once more than one context exists, every algorithm that constructs intermediate tensors context-free has the same failure mode. Stage 2 should therefore require that **all internal tensor creation goes through the caller's context** (`ctx.from_dense(...)`, `ctx.ones(...)`), making this class of bug unrepresentable. Placement-derived-from-tensors remains, but only as validation that inputs match the caller's context (typed error on mismatch), never as the mechanism by which fresh tensors pick a home. #720 should be treated as the first migration case of this rule rather than a one-off fix.

### Constraints imposed by hataori-rs

hataori-rs uses registered actions with no closure transfer, and `MPI_THREAD_FUNNELED`. Consequences for this design:

- A task payload can carry only host-serialized tensors plus a device designation (e.g. an ordinal). Execution contexts are constructed and owned on the worker and looked up by key, so the natural structure is a per-rank context table, not a process-global singleton.
- Serialization of device-resident values must be a typed error; `download` stays the only device-to-host boundary. This matches the existing "no implicit transfer" invariant.
- Task granularity is seconds to minutes, so context construction and session entry costs are negligible. A simple ownership model (independent context per task where convenient) beats clever cross-task cache sharing.

### API consequences for Stage 2

1. One `ExecutionContext` abstraction (enum over `Arc<CpuExecutionContext>` / `Arc<CudaExecutionContext>`, or a trait) accepted by every algorithm entry point: contract, zipup, SRC, fit, factorize.
2. All internal tensor creation is context-scoped (see above).
3. Prefer passing context values over closure-scoped sessions in the public API. The existing invariant "a high-level operation enters at most one backend session" already supports this: the public API takes `&ExecutionContext`, and each high-level operation enters its session internally exactly once. Closure nesting composes poorly with task bodies dispatched by an external scheduler.
4. `CudaExecutionContext::new_on(ordinal)`; document that mixing values from different CUDA contexts is a typed error (already partially covered by `validate_cuda_placement`).
5. `CpuExecutionContext::with_thread_pool(...)` for nested-parallelism control.

None of this changes Stage 1 sequencing; it changes what Stage 2 must look like and reframes the single-GPU restriction as temporary. Related: #720, #553.

**Comment by shinaoka** [comment]
Decision recorded in #720: implement placement/context-coherent CUDA SRC as the root fix. An early typed rejection may be an interim safety measure, but it does not satisfy #720. The implementation must use a reusable context-scoped construction/execution seam aligned with this roadmap, not an SRC-only CUDA shim.
